### PR TITLE
instanceof pattern matching produces redundant bytecode #585

### PR DESCRIPTION
**Pull request by trancexpress**
This change adjusts InstanceOfExpression.generateOptimizedBoolean() to not generate the redundant IF_ACMPNE byte code.

Fixes: #585

## Author checklist

- [ ] I have thoroughly tested my changes
- [ ] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [ ] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by trancexpress** [comment]
@iloveeclipse in #585 I see you have more than 1 question:

```
    ALOAD 4  // load "n" to stack AGAIN ???
    CHECKCAST Integer // checks "n" instanceof Number AGAIN ???
    IF_ACMPNE L2 // compares "i" and "n" and if not equal, returns ???
```

The current change only removes the `CHECKCAST`, i.e. the change is conservative. Do we want to see if we can remove `IF_ACMPNE` and the `ALOAD` as well? Assuming no tests fail due to the current change.

**Comment by trancexpress** [comment]
Seems to not be enough to fix the `spotbugs` issue.

**Comment by trancexpress** [comment]
The last change prevents the `spotbugs` issue. It doesn't deal with all the redundant byte code though. Lets see if tests are fine, if so we can discuss whether we invest more time here.

**Comment by iloveeclipse** [comment]
> Lets see if tests are fine, if so we can discuss whether we invest more time here.

Unfortunately not.

**Comment by trancexpress** [comment]
Looks like there are runtime errors now, even for simple pattern `instanceof` expressions. Likely we will have to "prune" all of the redundant code, when we have expressions without a nested `switch`...

**Comment by trancexpress** [comment]
Assuming I remove the entire block (rest stays the same):

```
    DUP
    ASTORE 2
    ALOAD 4
    CHECKCAST Integer
    IF_ACMPNE L2
```

I still get errors when running the generated byte code:

```
Error: Unable to initialize main class test.TestInstanceofPattern
Caused by: java.lang.VerifyError: Bad local variable type
Exception Details:
  Location:
    test/TestInstanceofPattern.main([Ljava/lang/String;)V @34: aload_2
  Reason:
    Type top (current frame, locals[2]) is not assignable to reference type
  Current Frame:
    bci: @34
    flags: { }
    locals: { '[Ljava/lang/String;', 'java/lang/Integer', top, top, 'java/lang/Integer' }
    stack: { 'java/lang/Integer', 'java/io/PrintStream', 'java/lang/StringBuilder' }
  Bytecode:
    0000000: 1210 b800 124c 2b3a 0419 04c1 0013 9900
    0000010: 1e19 04c0 0013 b200 18bb 001e 5912 20b7
    0000020: 0022 2cb6 0025 b600 29b6 002d b1       
  Stackmap Table:
    append_frame(@44,Object[#57])
```

**Comment by iloveeclipse** [comment]
Just wonder why do we need temporary variable at all? Does javac creates that too?

**Comment by trancexpress** [comment]
`org.eclipse.jdt.core.tests.compiler.regression.PatternMatching16Test` still fails, no idea how to run it from Eclipse (there are errors when I try).
`org.eclipse.jdt.core.tests.compiler.regression.RecordPatternTest` now also fails.

**Comment by jarthana** [comment]
> Just wonder why do we need temporary variable at all? Does javac creates that too?

It was introduced when we ran into stackmap frame issues. It was getting very difficult to manage values just within stack when branches were involved.

Tagging @mpalat

**Comment by trancexpress** [comment]
Seems like this doesn't work with the last change:

```
public static int foo(Object o) {
   return switch (o) {
        case Number n when (n instanceof Integer i) : {
            yield n.intValue() + i; // Error here
        }
        case Number n2 : {
            yield n2.intValue();
        }
        default : {
            yield -1;
        }
    };
}
```

So its not only a problem when nesting a `switch` in the pattern `instanceof`, its also a problem when the pattern `instanceof` is nested in a `switch`.

**Comment by trancexpress** [comment]
> > Just wonder why do we need temporary variable at all? Does javac creates that too?
> 
> It was introduced when we ran into stackmap frame issues. It was getting very difficult to manage values just within stack when branches were involved.
> 
> Tagging @mpalat

@jarthana , if you have time, can you explain *why* the `IF_ACMPNE` is needed here?

```
if (n instanceof Integer i)  {
```

```
    ALOAD 1: n
    ASTORE 4 // create the secret variable, set it to "n"
    ALOAD 4
    INSTANCEOF Integer
    IFEQ L2 // we should be done with the condition, but more code is generated
    ALOAD 4
    CHECKCAST Integer
    DUP
    ASTORE 2
    ALOAD 4
    CHECKCAST Integer
    IF_ACMPNE L2
   L3
    ... // if case
   L2
    ... // else case
```

Is it to somehow pop the secret variable from the stack? Since it seems to always compare the same value to itself (which is what `spotbugs` complains about).

When the `instanceof` is nested in a `switch`, the generated code is slightly different:

```
    LDC "zero"
   L2
    LINENUMBER 5 L2
    ASTORE 3
    ALOAD 3
    INSTANCEOF String
    IFEQ L3
    ALOAD 3
    CHECKCAST String
    DUP
    ASTORE 1
    ALOAD 3
    CHECKCAST String
   L4
    LINENUMBER 8 L4
    IF_ACMPNE L3
   L5
     ... // if case
   L3
     ... // else case
```

`L4` in this case maybe means some other code is meant to jump to the `IF_ACMPNE` and so for whichever case that is, maybe the `IF_ACMPNE` is not a noop?

**Comment by trancexpress** [comment]
Another bug in `ecj` with this snippet, not sure if there already is a `spotbugs` issue for it:

```
public class InstanceOfPatternTest {
	static String STR = "2";
	public static void main(String[] args) {
		if ( switch(STR) {
				default -> "zero";
			  }
				instanceof String s) {
			System.out.println(s);
		}
	}
}
```

```
Bug: Return value of String.hashCode() ignored in test.InstanceOfPatternTest.main(String[])
```
(at `if ( switch(STR) {`)

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java [hunk 1]
```diff
@@ -216,12 +216,14 @@ public void generateOptimizedBoolean(BlockScope currentScope, CodeStream codeStr
...
 				if (trueLabel != null) {
 					// Implicit falling through the FALSE case
-					codeStream.if_acmpeq(trueLabel);
+					//codeStream.if_acmpeq(trueLabel);
+					codeStream.pop2();
 				}
 			} else {
 				if (trueLabel == null) {
 					// Implicit falling through the TRUE case
-					codeStream.if_acmpne(falseLabel);
```
[the remark this answers is not in the document]
Comment (trancexpress) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java:224]: OK, looks like this breaks the following code:

```
	public static void main(String[] args) {
		System.out.println(foo(new Integer(10)));
		System.out.println(foo(new Integer(5)));
		System.out.println(foo(new Long(5L)));
		System.out.println(foo(new Float(0)));
	}
	public static int foo(Object o)  {
	   return switch (o) {
	        case Number n when (n instanceof Integer i) : {
	            yield n.intValue() + i; // Error here
	        }
	        case Number n2 : {
	            yield n2.intValue();
	        }
	        default : {
	            yield -1;
	        }
	    };
	}
```
(this requires JDK 19 on classpath and enabled Java 19 preview features)

```
org.eclipse.jdt.core.tests.compiler.regression.SwitchPatternTest.testBug578416_2 - 19
```

**Comment by jarthana** [comment]
> I honestly believe that removing this temporary variable and address the stack map frame problems would be the right way to fix this issue. I see the code is coming from https://bugs.eclipse.org/bugs/show_bug.cgi?id=562101 / https://git.eclipse.org/r/c/jdt/eclipse.jdt.core/+/166872/2/org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java. @mpalat : may be we can evaluate a different solution here?

If anything, the temporary variables make generating code easier. We have precedence in ForeachStatement and may be in other places too. 

Things can get really complicated when you mix case statements with instanceof patterns. A different solution would probably mean ripping apart the code generation of the relevant nodes. Considering the IOE is a subclass of BinaryExpression, we might have serious consequences. We can try, but won't be a trivial exercise, IMO.

**Comment by trancexpress** [comment]
Hi @jarthana , @mpalat ,

please take a look here. From what I see, the generated byte code loads the same value twice then does a comparison and jumps based on the result.

The current change replaces the current comparison with a `pop2`, since the comparison does nothing and a `goto` in the case of using `if_acmpeq`. This seems to make all tests happy (aside from 1 test with hard-coded expectations of generated byte code; that snippet executes as expected with the change).

This would be bare minimum to make `spotbugs` happy and to not generate redundant code (to an extent). Can you check the change and let us know of your opinions? I'll need to remove the 2 commented out lines in the final change.

Ideally we clean more of the generated code up, since loading the same value twice, checking it if its of a specific type twice, and then doing a noop compare (or pops, with the current change) is not nice. Do you think we should try to clean up more, or we should leave the change at its current minimal state? Is this code known to be fragile, or with insufficient tests?

Best regards and thanks,
Simeon

**Comment by jarthana** [comment]
> Hi @jarthana , @mpalat ,
> 
> please take a look here. From what I see, the generated byte code loads the same value twice then does a comparison and jumps based on the result.

Manoj is off and will only be back on Monday. Hopefully he can spare some time for this soon.

Since we are still in preview phase of the new record pattern feature as well as pattern matching for switch, any further cleanup will be welcome. If there are concerns of breaking, then we can even put those changes in the beta branches, so that we get some breathing space.

**Comment by trancexpress** [comment]
@iloveeclipse , should I remove the commented out lines and rebase the commit?

**Comment by iloveeclipse** [comment]
> @iloveeclipse , should I remove the commented out lines and rebase the commit?

Yes, but please answer on my comment above first (missing goto?).

**Comment by trancexpress** [comment]
> Yes, but please answer on my comment above first (missing goto?).

I already did? Should I resolve the conversation?

**Comment by iloveeclipse** [comment]
> > Yes, but please answer on my comment above first (missing goto?).
> 
> I already did? Should I resolve the conversation?

Sorry, I don't see that.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java [hunk 2]
```diff
@@ -216,12 +216,15 @@ public void generateOptimizedBoolean(BlockScope currentScope, CodeStream codeStr
...
+					//codeStream.if_acmpeq(trueLabel);
+					codeStream.pop2();
+					codeStream.goto_(trueLabel);
 				}
 			} else {
 				if (trueLabel == null) {
 					// Implicit falling through the TRUE case
-					codeStream.if_acmpne(falseLabel);
+					//codeStream.if_acmpne(falseLabel);
+					codeStream.pop2();
```
[the remark this answers is not in the document]
Comment (iloveeclipse) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java:225]: This comment has no answer yet

**Comment by trancexpress** [comment]
> Sorry, I don't see that.

The 2 values that are compared are always equal, so the code never jumps. Adding a jump will break snippets.

**Pull request merged**